### ISSUE TITLE
[FEA] Port NVStrings split_record/rsplit_record to libcudf++

### ISSUE THREAD

**Issue by davidwendt** [rapidsai/cudf#3825]
This is to cover porting of the `split_record` and `rsplit_record` functions in `NVStrings` to `libcudf++`.

`NVStrings` documentation for split_record: https://rapidsai.github.io/projects/libnvstrings/0.11.0/classNVStrings.html#a9c7936e82eda42aa477fcd91e472ffe7
`NVStrings` documentation for rsplit_record: https://rapidsai.github.io/projects/libnvstrings/0.11.0/classNVStrings.html#ab4ee93a3dd55f274096e203d7653d86b

There are actually 4 functions to port. 2 for `split_record` and 2 for `rsplit_record`. The split logic for characters vs whitespace is very different and therefore were divided into separate functions.

I expect these to follow closely to the `split()` and `rsplit()` counterparts defined in https://github.com/rapidsai/cudf/blob/branch-0.12/cpp/include/cudf/strings/split/split.hpp

Suggested API signature:
```
std::vector<std::unique_ptr<column>> split_record( strings_column_view const& strings,
                                            string_scalar const& delimiter = string_scalar(""),
                                            size_type maxsplit=-1,
                                            rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());

std::vector<std::unique_ptr<column>> rsplit_record( strings_column_view const& strings,
                                            string_scalar const& delimiter = string_scalar(""),
                                            size_type maxsplit=-1,
                                            rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());

```
These cannot return a `table` since the individual columns may be of different lengths. It may be possible to refactor and reuse some of the functors from the existing `split()/rsplit()` implementation.

These `_record` versions will split strings _horizontally_. This means that each string from the input column will create a column of new strings. If the input column has 1000 strings, the result should be 1000 columns. Therefore, I recommend looking at `contiguous_split` to see how they create a single allocate for multiple columns.

**Comment by seunghwak** [comment]
@davidwendt 
If I am not mistaken, if I follow the contiguous_split's approach,

https://github.com/rapidsai/cudf/blob/07a755a569e604cbb2924c385f017a610d6f01d6/cpp/include/cudf/copying.hpp#L435

I should define

```
struct contiguous_split_record_result {
  std::vector<column_view>
  std::unique_ptr<rmm::device_buffer> all_data
}
```

and the public API should look like

```
contiguous_split_record_result
contiguous_split_record(strings_column_view& strings, ...)

contiguous_split_record_result
contiguous_split_record(strings_column_view& strings, ...)
```

This differs from the suggested API signature above. Any thoughts on this?

Or should we define both contiguous_split_record (returning contiguous_split_record_result) and split_record (returning std::vector<std::unique_ptr<column>> possibly requiring a large number of memory allocations)?

And is this function for strings only?

**Comment by davidwendt** [comment]
Yes, this looks correct to me since you need to pass the overall `device_buffer` as well the vector of `column_view`s.
And yes, this is a strings-only function. We are splitting the actual string characters up within each string to make new strings. There is no fixed-width equivalent of this type of operation.

**Comment by harrism** [comment]
Fixed by #3891
